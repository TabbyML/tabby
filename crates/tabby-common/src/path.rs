use std::{cell::Cell, env, path::PathBuf, sync::Mutex};

use lazy_static::lazy_static;

lazy_static! {
    static ref TABBY_ROOT: Mutex<Cell<PathBuf>> = {
        Mutex::new(Cell::new(match env::var("TABBY_ROOT") {
            Ok(x) => PathBuf::from(x),
            Err(_) => home::home_dir().unwrap().join(".tabby"),
        }))
    };
    static ref TABBY_MODEL_CACHE_ROOT: Option<PathBuf> =
        env::var("TABBY_MODEL_CACHE_ROOT").ok().map(PathBuf::from);
}

#[cfg(any(feature = "testutils", test))]
pub fn set_tabby_root(path: PathBuf) {
    println!("SET TABBY ROOT: '{}'", path.display());
    let cell = TABBY_ROOT.lock().unwrap();
    cell.replace(path);
}

pub fn tabby_root() -> PathBuf {
    let mut cell = TABBY_ROOT.lock().unwrap();
    cell.get_mut().clone()
}

pub fn config_file() -> PathBuf {
    tabby_root().join("config.toml")
}

pub fn usage_id_file() -> PathBuf {
    tabby_root().join("usage_anonymous_id")
}

pub fn repositories_dir() -> PathBuf {
    tabby_root().join("repositories")
}

pub fn index_dir() -> PathBuf {
    tabby_root().join("index")
}

pub fn models_dir() -> PathBuf {
    if let Some(cache_root) = &*TABBY_MODEL_CACHE_ROOT {
        cache_root.clone()
    } else {
        tabby_root().join("models")
    }
}

pub fn events_dir() -> PathBuf {
    tabby_root().join("events")
}

mod registry {}
