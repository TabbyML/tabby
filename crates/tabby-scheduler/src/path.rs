use std::path::PathBuf;

use tabby_common::path::tabby_root;

pub fn incremental_repository_store() -> PathBuf {
    tabby_root().join("incremental-repositorries.kv")
}
