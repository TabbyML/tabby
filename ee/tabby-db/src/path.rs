use std::path::PathBuf;

use tabby_common::path::tabby_root;

fn tabby_ee_root() -> PathBuf {
    tabby_root().join("ee")
}

pub fn db_file() -> PathBuf {
    #[cfg(feature = "prod")]
    {
        tabby_ee_root().join("db.sqlite")
    }
    #[cfg(not(feature = "prod"))]
    {
        tabby_ee_root().join("dev-db.sqlite")
    }
}
