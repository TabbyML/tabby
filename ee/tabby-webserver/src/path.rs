use std::path::PathBuf;

use tabby_common::path::tabby_root;

pub fn tabby_ee_root() -> PathBuf {
    tabby_root().join("ee")
}

pub fn db_file() -> PathBuf {
    #[cfg(feature = "prod-db")]
    {
        tabby_ee_root().join("db.sqlite")
    }
    #[cfg(not(feature = "prod-db"))]
    {
        tabby_ee_root().join("dev-db.sqlite")
    }
}

pub fn job_db_file() -> PathBuf {
    tabby_ee_root().join("job.sqlite")
}
