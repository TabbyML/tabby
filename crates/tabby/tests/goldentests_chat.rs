use std::path::PathBuf;

use futures::StreamExt;
use lazy_static::lazy_static;
use reqwest_eventsource::{Event, EventSource};
use serde::Deserialize;
use serde_json::json;
use serial_test::serial;
use tokio::{
    process::Command,
    time::{sleep, Duration},
};

#[derive(Deserialize)]
pub struct ChatCompletionChunk {
    choices: [ChatCompletionChoice; 1],
}

#[derive(Deserialize)]
pub struct ChatCompletionChoice {
    delta: ChatCompletionDelta,
}

#[derive(Deserialize)]
pub struct ChatCompletionDelta {
    content: String,
}

lazy_static! {
    static ref CLIENT: reqwest::Client = reqwest::Client::new();
}

fn workspace_dir() -> PathBuf {
    let output = std::process::Command::new(env!("CARGO"))
        .arg("locate-project")
        .arg("--workspace")
        .arg("--message-format=plain")
        .output()
        .unwrap()
        .stdout;
    let cargo_path = std::path::Path::new(std::str::from_utf8(&output).unwrap().trim());
    cargo_path.parent().unwrap().to_path_buf()
}

fn tabby_path() -> PathBuf {
    workspace_dir().join("target/debug/tabby")
}

fn initialize_server(gpu_device: Option<&str>) {
    let mut cmd = Command::new(tabby_path());
    cmd.arg("serve")
        .arg("--chat-model")
        .arg("TabbyML/Mistral-7B")
        .arg("--no-webserver")
        .arg("--port")
        .arg("9090")
        .kill_on_drop(true);

    if let Some(gpu_device) = gpu_device {
        cmd.arg("--device").arg(gpu_device);
    }

    tokio::task::spawn(async move {
        cmd.spawn()
            .expect("Failed to start server")
            .wait()
            .await
            .unwrap();
    });
}

async fn wait_for_server(gpu_device: Option<&str>) {
    initialize_server(gpu_device);

    loop {
        println!("Waiting for server to start...");
        let is_ok = reqwest::get("http://127.0.0.1:9090/v1/health")
            .await
            .is_ok();
        if is_ok {
            break;
        } else {
            sleep(Duration::from_secs(5)).await;
        }
    }
}

async fn golden_test(body: serde_json::Value) -> String {
    let mut es = EventSource::new(
        CLIENT
            .post("http://127.0.0.1:9090/v1/chat/completions")
            .json(&body),
    )
    .unwrap();

    let mut actual = "".to_owned();
    while let Some(event) = es.next().await {
        match event {
            Ok(Event::Open) => {}
            Ok(Event::Message(message)) => {
                let x: ChatCompletionChunk = serde_json::from_str(&message.data).unwrap();
                actual += &x.choices[0].delta.content;
            }
            Err(_) => {
                // StreamEnd
                break;
            }
        }
    }

    actual
}

macro_rules! assert_golden {
    ($expr:expr) => {
        insta::assert_yaml_snapshot!(golden_test($expr).await);
    };
}

#[cfg(all(target_os = "macos", target_arch = "aarch64"))]
#[tokio::test]
#[serial]
async fn run_chat_golden_tests() {
    wait_for_server(Some("metal")).await;

    assert_golden!(json!({
            "seed": 0,
            "messages": [
                {
                    "role": "user",
                    "content": "How to convert a list of string to numbers in python"
                }
            ]
    }));

    assert_golden!(json!({
            "seed": 0,
            "messages": [
                {
                    "role": "user",
                    "content": "How to parse email address with regex"
                }
            ]
    }));
}

#[tokio::test]
#[serial]
async fn run_chat_golden_tests_cpu() {
    wait_for_server(Some("cpu")).await;

    assert_golden!(json!({
            "seed": 0,
            "messages": [
                {
                    "role": "user",
                    "content": "How are you?"
                }
            ]
    }));
}
