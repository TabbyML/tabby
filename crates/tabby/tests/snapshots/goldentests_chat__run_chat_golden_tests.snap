---
source: crates/tabby/tests/goldentests_chat.rs
expression: "golden_test(json!({\n                \"seed\": 0, \"messages\":\n                [{\n                    \"role\": \"user\", \"content\":\n                    \"How to convert a list of string to numbers in python\"\n                }]\n            })).await"
---
" In Python, you can convert a list of strings to numbers using the `map()` function and the `int()` function. Here's an example:\n```\nstrings = ['1', '2', '3', '4', '5']\nnumbers = list(map(int, strings))\nprint(numbers)\n```\nThis will output:\n```\n[1, 2, 3, 4, 5]\n```\nIn this example, the `map()` function applies the `int()` function to each element of the `strings` list, converting each string to an integer and returning a new list of integers. The `list()` function is used to convert the result of the `map()` function back into a list."

